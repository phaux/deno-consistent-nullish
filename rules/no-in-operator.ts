import { fc } from "../util/formatCode.ts";
import { getOperatorRange } from "../util/getOperatorRange.ts";

export const noInOperatorRule: Deno.lint.Rule = {
  create: (ctx) => ({
    BinaryExpression(exprNode) {
      if (exprNode.operator !== "in") return;

      let newPropText = `[${ctx.sourceCode.getText(exprNode.left)}]`;
      if (
        exprNode.left.type === "Literal" &&
        typeof exprNode.left.value === "string" &&
        exprNode.left.value.match(/^\w[\w\d]*$/)
      ) {
        newPropText = `.${exprNode.left.value}`;
      }

      ctx.report({
        range: getOperatorRange(ctx.sourceCode, exprNode),
        message: "`in` operator is not allowed.",
        hint:
          "`in` operator differentiates between missing property and a property with nullish value. " +
          `Use ${fc(`obj${newPropText} != null`)} to check for missing or nullish property.`,
        *fix(fixer) {
          // Remove left operand and `in` operator.
          yield fixer.removeRange([
            exprNode.left.range[0],
            exprNode.right.range[0],
          ]);
          // Add property access.
          yield fixer.insertTextAfterRange(
            exprNode.right.range,
            `${newPropText} != null`,
          );
        },
      });
    },
  }),
};
