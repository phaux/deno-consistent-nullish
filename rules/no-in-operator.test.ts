import { assertSnapshot } from "@std/testing/snapshot";
import { noInOperatorRule } from "./no-in-operator.ts";
import { assertEquals } from "@std/assert";

Deno.test("bans in operator", async (t) => {
  const diagnostics = Deno.lint.runPlugin(
    { name: "test", rules: { rule: noInOperatorRule } },
    "test.ts",
    `
    if (x in obj) {}
    if ("foo" in obj) {}
    if (!("foo" in obj)) {}
    if ("a" + "b" in [] + []) {}
    if ("foo" in getObj()) {}
    if (1 in obj) {}
    if (x in {}) {}
    `
  );
  assertEquals(diagnostics.length, 7);
  await assertSnapshot(t, diagnostics);
});
