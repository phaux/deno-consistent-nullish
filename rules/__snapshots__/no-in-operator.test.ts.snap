export const snapshot = {};

snapshot[`bans in operator 1`] = `
[
  {
    fix: [
      {
        range: [
          9,
          14,
        ],
        text: "",
      },
      {
        range: [
          17,
          17,
        ],
        text: "[x] != null",
      },
    ],
    hint: "\`in\` operator differentiates between missing property and a property with nullish value. Use \`obj[x] != null\` to check for missing or nullish property.",
    id: "test/rule",
    message: "\`in\` operator is not allowed.",
    range: [
      11,
      13,
    ],
  },
  {
    fix: [
      {
        range: [
          30,
          39,
        ],
        text: "",
      },
      {
        range: [
          42,
          42,
        ],
        text: ".foo != null",
      },
    ],
    hint: "\`in\` operator differentiates between missing property and a property with nullish value. Use \`obj.foo != null\` to check for missing or nullish property.",
    id: "test/rule",
    message: "\`in\` operator is not allowed.",
    range: [
      36,
      38,
    ],
  },
  {
    fix: [
      {
        range: [
          57,
          66,
        ],
        text: "",
      },
      {
        range: [
          69,
          69,
        ],
        text: ".foo != null",
      },
    ],
    hint: "\`in\` operator differentiates between missing property and a property with nullish value. Use \`obj.foo != null\` to check for missing or nullish property.",
    id: "test/rule",
    message: "\`in\` operator is not allowed.",
    range: [
      63,
      65,
    ],
  },
  {
    fix: [
      {
        range: [
          83,
          96,
        ],
        text: "",
      },
      {
        range: [
          103,
          103,
        ],
        text: '["a" + "b"] != null',
      },
    ],
    hint: '\`in\` operator differentiates between missing property and a property with nullish value. Use \`obj["a" + "b"] != null\` to check for missing or nullish property.',
    id: "test/rule",
    message: "\`in\` operator is not allowed.",
    range: [
      93,
      95,
    ],
  },
  {
    fix: [
      {
        range: [
          116,
          125,
        ],
        text: "",
      },
      {
        range: [
          133,
          133,
        ],
        text: ".foo != null",
      },
    ],
    hint: "\`in\` operator differentiates between missing property and a property with nullish value. Use \`obj.foo != null\` to check for missing or nullish property.",
    id: "test/rule",
    message: "\`in\` operator is not allowed.",
    range: [
      122,
      124,
    ],
  },
  {
    fix: [
      {
        range: [
          146,
          151,
        ],
        text: "",
      },
      {
        range: [
          154,
          154,
        ],
        text: "[1] != null",
      },
    ],
    hint: "\`in\` operator differentiates between missing property and a property with nullish value. Use \`obj[1] != null\` to check for missing or nullish property.",
    id: "test/rule",
    message: "\`in\` operator is not allowed.",
    range: [
      148,
      150,
    ],
  },
  {
    fix: [
      {
        range: [
          167,
          172,
        ],
        text: "",
      },
      {
        range: [
          174,
          174,
        ],
        text: "[x] != null",
      },
    ],
    hint: "\`in\` operator differentiates between missing property and a property with nullish value. Use \`obj[x] != null\` to check for missing or nullish property.",
    id: "test/rule",
    message: "\`in\` operator is not allowed.",
    range: [
      169,
      171,
    ],
  },
]
`;
